File: validator-b/spd_validator_b/passive.py
"""Independent namespace/token-based passive inspection. Never fetch or rewrite."""
from __future__ import annotations

import posixpath
import unicodedata
from urllib.parse import unquote, urlsplit

import tinycss2
from lxml import etree

from .models import Outcome, ResourceLimitExceeded
from .xmlutil import XMLValidationError, parse_xml

SVG = 'http://www.w3.org/2000/svg'
XHTML = 'http://www.w3.org/1999/xhtml'
MATH = 'http://www.w3.org/1998/Math/MathML'


def url_value(value):
    return value.strip(' \t\r\n\f').replace('\t', '').replace('\r', '').replace('\n', '')


class Inspector:
    def __init__(self, package, report):
        self.package, self.report = package, report
        self.xml = {}
        self.edges = {}

    def fail(self, path, *ids, evidence=''):
        for rid in ids:
            self.report.add(rid, Outcome.FAIL, 'passive content/profile violation', resource=path, evidence=evidence)

    def reference(self, path, value, role='other', svg=False, hyperlink=False, source=None):
        value = url_value(value)
        try:
            parts = urlsplit(value)
        except ValueError:
            self.fail(path, 'SPD-SEC-008', evidence=value)
            return
        scheme = parts.scheme.lower()
        if hyperlink and scheme in {'http', 'https'} and parts.netloc:
            return
        ids = set()
        if scheme in {'javascript', 'vbscript'}:
            ids.add('SPD-SEC-001')
        forbidden = bool(scheme or parts.netloc or value.startswith(('/', '\\')))
        if forbidden:
            ids.add('SPD-SEC-008')
            if not hyperlink:
                ids.add('SPD-SEC-003')
                specific = {'font': 'SPD-SEC-004', 'stylesheet': 'SPD-SEC-005', 'image': 'SPD-SEC-006'}.get(role)
                if specific:
                    ids.add(specific)
            if svg:
                ids.add('SPD-SEC-007')
            self.fail(path, *sorted(ids), evidence=value)
            return
        try:
            decoded = unquote(parts.path, errors='strict')
            fragment = unquote(parts.fragment, errors='strict')
        except (ValueError, UnicodeDecodeError):
            self.fail(path, 'SPD-SEC-008', evidence=value)
            return
        target = posixpath.normpath(posixpath.join(posixpath.dirname(path), decoded)) if decoded else path
        unsafe = parts.query or '\\' in decoded or ':' in decoded or decoded.startswith('/') or target.startswith('../') or target == '..' or unicodedata.normalize('NFC', target) != target
        if unsafe or not self.package.exists(target):
            self.fail(path, 'SPD-SEC-008', evidence=value)
            return
        if fragment and (target not in self.xml or not self.xml[target].xpath('//*[@id=$id]', id=fragment)):
            self.fail(path, 'SPD-SEC-008', evidence=value)
        if not hyperlink and source is not None:
            self.edges.setdefault(source, set()).add(target + ('#' + fragment if fragment else ''))

    def css(self, path, text, svg=False):
        # Component-value tokens preserve nested blocks/functions and decode escapes.
        def walk(tokens, role='image', depth=0, strings=False):
            if depth > 128:
                raise ResourceLimitExceeded('CSS nesting inspection budget exceeded')
            import_pending = False
            font_pending = False
            for token in tokens:
                typ = token.type
                if typ in {'whitespace', 'comment'}:
                    continue
                if typ == 'at-keyword':
                    import_pending = token.lower_value == 'import'
                    font_pending = token.lower_value == 'font-face'
                elif typ == 'ident' and token.lower_value in {'behavior', '-moz-binding'}:
                    self.fail(path, 'SPD-SEC-001')
                elif typ == 'url' or typ == 'string' and (import_pending or strings):
                    self.reference(path, token.value, 'stylesheet' if import_pending else role, svg, source=path if import_pending else None)
                    import_pending = False
                elif typ == 'function':
                    name = token.lower_name
                    if strings and name in {'var', 'attr', 'env'}:
                        self.report.add('SPD-SEC-008', Outcome.NOT_TESTED, 'dynamic string resource consumer not resolved', resource=path)
                    if name in {'url', 'src'}:
                        args = [x for x in token.arguments if x.type not in {'whitespace', 'comment'}]
                        if len(args) == 1 and args[0].type in {'string', 'url'}:
                            self.reference(path, args[0].value, 'stylesheet' if import_pending else role, svg, source=path if import_pending else None)
                        else:
                            self.report.add('SPD-SEC-008', Outcome.NOT_TESTED, 'dynamic URL consumer requires unresolved substitution evaluation', resource=path)
                        import_pending = False
                    elif name == 'local' and role == 'font':
                        self.fail(path, 'SPD-SEC-004')
                    elif name == 'expression':
                        self.fail(path, 'SPD-SEC-001')
                    else:
                        walk(token.arguments, role, depth + 1, name in {'image', 'image-set', '-webkit-image-set'})
                elif hasattr(token, 'content'):
                    walk(token.content, 'font' if font_pending else role, depth + 1)
                    font_pending = False
                    import_pending = False
                elif typ == 'error':
                    self.fail(path, 'SPD-SEC-005', evidence=token.message)
                elif typ == 'literal' and token.value == ';':
                    import_pending = False
                    font_pending = False
        walk(tinycss2.parse_component_value_list(text), depth=0)

    def inspect_xml(self, path, root):
        for pi in root.getroottree().xpath('//processing-instruction("xml-stylesheet")'):
            try:
                attrs = etree.fromstring(('<style ' + pi.text + '/>').encode()).attrib
                if attrs.get('href'):
                    self.reference(path, attrs['href'], 'stylesheet')
            except etree.XMLSyntaxError:
                self.fail(path, 'SPD-SEC-005')
        for element in root.iter():
            if not isinstance(element.tag, str):
                continue
            q = etree.QName(element)
            name = q.localname.lower()
            if q.namespace not in {SVG, XHTML, MATH}:
                continue
            svg = q.namespace == SVG or any(etree.QName(a).namespace == SVG for a in element.iterancestors())
            if name == 'script':
                self.fail(path, 'SPD-SEC-001', *(['SPD-SEC-007'] if svg else []))
            if name in {'iframe', 'object', 'embed', 'applet'}:
                self.fail(path, 'SPD-SEC-002', 'SPD-PASS-002')
            if name in {'form', 'base'} or element.get('{http://www.w3.org/XML/1998/namespace}base') is not None:
                self.fail(path, 'SPD-PASS-002')
            if name == 'foreignobject' and svg:
                self.fail(path, 'SPD-SEC-007', 'SPD-PASS-002')
            if name == 'canvas':
                self.fail(path, 'SPD-PASS-002')
            if name in {'input', 'button'}:
                label = element.get('aria-label') or element.get('aria-labelledby') or (''.join(element.itertext()).strip() if name == 'button' else None)
                if not label and element.get('id'):
                    label = root.xpath('//*[local-name()="label"][@for=$id]', id=element.get('id'))
                forbidden = {'form', 'formaction', 'formenctype', 'formmethod', 'formtarget', 'formnovalidate', 'command', 'commandfor', 'popovertarget', 'popovertargetaction'}
                bad = element.get('disabled') is None or not label or bool(forbidden.intersection(element.attrib))
                bad |= element.get('type', '').lower() in {'password', 'file', 'submit', 'image', 'reset'}
                bad |= name == 'button' and element.get('type', 'submit').lower() != 'button'
                bad |= name == 'input' and element.get('value') is None and element.get('type') not in {'checkbox', 'radio'}
                if bad:
                    self.fail(path, 'SPD-PASS-002')
            if name == 'meta' and element.get('http-equiv', '').lower() == 'refresh':
                self.fail(path, 'SPD-SEC-003', 'SPD-PASS-002')
            if 'autoplay' in element.attrib:
                self.fail(path, 'SPD-PASS-003')
            rel = set(element.get('rel', '').lower().split())
            if 'ping' in element.attrib or rel.intersection({'prefetch', 'preconnect', 'dns-prefetch', 'prerender'}):
                self.fail(path, 'SPD-SEC-003')
            if name == 'style':
                self.css(path, ''.join(element.itertext()), svg)
            for attr, value in element.attrib.items():
                local = etree.QName(attr).localname.lower()
                if local.startswith('on'):
                    self.fail(path, 'SPD-SEC-001', *(['SPD-SEC-007'] if svg else []))
                if local == 'style' or (svg and local in {'fill', 'stroke', 'filter', 'clip-path', 'mask', 'cursor', 'marker-start', 'marker-mid', 'marker-end'}):
                    self.css(path, value, svg)
                if local not in {'href', 'src', 'srcset', 'poster', 'data', 'background'}:
                    continue
                if name == 'base':
                    continue  # Prohibited override is not a rendered resource.
                hyperlink = local == 'href' and (name in {'a', 'area'} or (q.namespace == MATH and name != 'mglyph'))
                role = 'stylesheet' if name == 'link' and 'stylesheet' in rel else 'image' if name in {'img', 'image', 'use', 'feimage', 'mglyph', 'audio', 'video', 'source', 'track'} or local in {'poster', 'background'} else 'other'
                if name == 'link' and rel.intersection({'preload', 'modulepreload'}):
                    role = {'font': 'font', 'style': 'stylesheet', 'image': 'image'}.get(element.get('as'), role)
                values = [part.strip().split()[0] for part in value.split(',') if part.strip()] if local == 'srcset' else [value]
                for candidate in values:
                    source = None
                    if svg and not hyperlink:
                        owner = next((a.get('id') for a in [element, *element.iterancestors()] if a.get('id')), None)
                        source = path + ('#' + owner if owner else '')
                    self.reference(path, candidate, role, svg, hyperlink, source)

    def run(self):
        css_paths = set()
        # Parse content by namespace, not filename or embedding assumptions.
        for path in sorted(self.package.paths):
            if path.endswith('.css'):
                css_paths.add(path)
            data = self.package.read(path)
            if not (data.lstrip().startswith(b'<') or data.startswith((b'\xef\xbb\xbf', b'\xff\xfe', b'\xfe\xff', b'\x00<', b'<\x00'))):
                continue
            if len(data) > self.package.limits.max_xml_bytes:
                raise ResourceLimitExceeded('XML inspection byte budget exceeded')
            try:
                root = parse_xml(data, resource=path)
            except XMLValidationError:
                continue  # Existing XML/EPUB rules own malformed documents.
            self.xml[path] = root
            if etree.QName(root).namespace == 'http://www.idpf.org/2007/opf':
                for item in root.xpath('//*[local-name()="item"]'):
                    if item.get('media-type', '').lower() in {'application/javascript', 'text/javascript', 'application/ecmascript', 'text/ecmascript'}:
                        self.fail(path, 'SPD-SEC-001')
                for item in root.xpath('//*[local-name()="item"][@media-type="text/css"]'):
                    href = item.get('href', '')
                    if not urlsplit(href).scheme:
                        css_paths.add(posixpath.normpath(posixpath.join(posixpath.dirname(path), unquote(href))))
        for path, root in self.xml.items():
            self.inspect_xml(path, root)
        for path in sorted(css_paths & self.package.paths):
            try:
                self.css(path, self.package.read(path).decode('utf-8-sig'))
            except UnicodeDecodeError:
                self.fail(path, 'SPD-SEC-005')
        # Iterative graph traversal bounds stack use; hyperlink edges are excluded.
        done = set()
        for root in self.edges:
            pending = [(root, False)]
            active = set()
            while pending:
                node, leave = pending.pop()
                if leave:
                    active.discard(node)
                    done.add(node)
                elif node in active:
                    self.fail(root.split('#')[0], 'SPD-SEC-008', evidence='resource cycle')
                elif node not in done:
                    active.add(node)
                    pending.append((node, True))
                    pending.extend((child, False) for child in self.edges.get(node, ()))


def validate_passive(package, report):
    Inspector(package, report).run()

File: validator-a/src/passive.rs
//! Passive resource inspection using CSS Syntax tokens and namespace-aware XML.
use crate::{Finding, Outcome, Severity, package::Entry};
use cssparser::{Parser, ParserInput, Token};
use percent_encoding::percent_decode_str;
use roxmltree::Document;
use std::collections::{BTreeMap, BTreeSet};
use unicode_normalization::UnicodeNormalization;

const SVG: &str = "http://www.w3.org/2000/svg";
const HTML: &str = "http://www.w3.org/1999/xhtml";
const MATH: &str = "http://www.w3.org/1998/Math/MathML";

#[derive(Default)]
struct CssFacts {
    urls: Vec<(String, &'static str, bool)>,
    code: bool,
    local_font: bool,
    unresolved: bool,
    malformed: bool,
    limited: bool,
}

fn tokens(
    input: &mut Parser<'_, '_>,
    role: &'static str,
    strings: bool,
    depth: usize,
    out: &mut CssFacts,
) {
    if depth > 128 {
        out.limited = true;
        return;
    }
    let mut import = false;
    let mut font = false;
    while let Ok(token) = input.next_including_whitespace_and_comments().cloned() {
        match token {
            Token::AtKeyword(name) => {
                import = name.eq_ignore_ascii_case("import");
                font = name.eq_ignore_ascii_case("font-face");
            }
            Token::UnquotedUrl(value) => {
                out.urls.push((
                    value.to_string(),
                    if import { "stylesheet" } else { role },
                    import,
                ));
                import = false;
            }
            Token::QuotedString(value) if import || strings => {
                out.urls.push((
                    value.to_string(),
                    if import { "stylesheet" } else { role },
                    import,
                ));
                import = false;
            }
            Token::Ident(name)
                if matches!(
                    name.to_ascii_lowercase().as_str(),
                    "behavior" | "-moz-binding"
                ) =>
            {
                out.code = true
            }
            Token::Function(name) => {
                let name = name.to_ascii_lowercase();
                if strings && matches!(name.as_str(), "var" | "attr" | "env") {
                    out.unresolved = true;
                }
                let _: Result<(), cssparser::ParseError<'_, ()>> =
                    input.parse_nested_block(|nested| {
                        match name.as_str() {
                            "url" | "src" => {
                                if let Ok(value) = nested.expect_string_cloned() {
                                    if nested.is_exhausted() {
                                        out.urls.push((
                                            value.to_string(),
                                            if import { "stylesheet" } else { role },
                                            import,
                                        ));
                                    } else {
                                        out.unresolved = true;
                                    }
                                } else {
                                    out.unresolved = true;
                                }
                            }
                            "local" if role == "font" => out.local_font = true,
                            "expression" => out.code = true,
                            _ => tokens(
                                nested,
                                role,
                                matches!(
                                    name.as_str(),
                                    "image" | "image-set" | "-webkit-image-set"
                                ),
                                depth + 1,
                                out,
                            ),
                        }
                        Ok(())
                    });
                import = false;
            }
            Token::CurlyBracketBlock | Token::SquareBracketBlock | Token::ParenthesisBlock => {
                let _: Result<(), cssparser::ParseError<'_, ()>> =
                    input.parse_nested_block(|nested| {
                        tokens(
                            nested,
                            if font { "font" } else { role },
                            false,
                            depth + 1,
                            out,
                        );
                        Ok(())
                    });
                font = false;
                import = false;
            }
            Token::BadUrl(_) | Token::BadString(_) => out.malformed = true,
            Token::Semicolon => {
                import = false;
                font = false;
            }
            _ => {}
        }
    }
}

struct Inspection<'a> {
    entries: &'a BTreeMap<String, Entry>,
    documents: BTreeMap<String, Document<'a>>,
    findings: &'a mut Vec<Finding>,
    edges: BTreeMap<String, BTreeSet<String>>,
}

impl Inspection<'_> {
    fn issue(&mut self, path: &str, id: &str, message: &str, outcome: Outcome) {
        self.findings.push(Finding {
            requirement_id: id.into(),
            outcome,
            severity: if outcome == Outcome::Fail {
                Severity::Error
            } else {
                Severity::NotTested
            },
            capability: "Base".into(),
            stage: "security".into(),
            resource: Some(path.into()),
            node_id: None,
            revision_id: None,
            message: message.into(),
            evidence: serde_json::Value::Null,
            source: "Validator A".into(),
        });
    }
    fn fail(&mut self, path: &str, id: &str) {
        self.issue(path, id, "passive profile violation", Outcome::Fail);
    }

    fn reference(
        &mut self,
        path: &str,
        raw: &str,
        role: &str,
        svg: bool,
        link: bool,
        source: Option<String>,
    ) {
        let value = raw
            .trim_matches([' ', '\t', '\r', '\n', '\u{c}'])
            .replace(['\t', '\r', '\n'], "");
        let scheme = value
            .split_once(':')
            .map(|(s, _)| s.to_ascii_lowercase())
            .unwrap_or_default();
        if link && matches!(scheme.as_str(), "http" | "https") && value.contains("://") {
            return;
        }
        if !scheme.is_empty() || value.starts_with(['/', '\\']) {
            if matches!(scheme.as_str(), "javascript" | "vbscript") {
                self.fail(path, "SPD-SEC-001");
            }
            self.fail(path, "SPD-SEC-008");
            if !link {
                self.fail(path, "SPD-SEC-003");
                match role {
                    "font" => self.fail(path, "SPD-SEC-004"),
                    "stylesheet" => self.fail(path, "SPD-SEC-005"),
                    "image" => self.fail(path, "SPD-SEC-006"),
                    _ => {}
                }
            }
            if svg {
                self.fail(path, "SPD-SEC-007");
            }
            return;
        }
        let (raw_path, fragment) = value.split_once('#').unwrap_or((&value, ""));
        let Ok(decoded) = percent_decode_str(raw_path).decode_utf8() else {
            self.fail(path, "SPD-SEC-008");
            return;
        };
        let mut parts: Vec<&str> = if decoded.is_empty() {
            path.split('/').collect()
        } else {
            path.rsplit_once('/')
                .map(|(p, _)| p.split('/').collect())
                .unwrap_or_default()
        };
        let mut unsafe_path = decoded.starts_with('/') || decoded.contains(['\\', ':', '?']);
        if !decoded.is_empty() {
            for part in decoded.split('/') {
                match part {
                    "." | "" => {}
                    ".." => {
                        if parts.pop().is_none() {
                            unsafe_path = true;
                        }
                    }
                    p => parts.push(p),
                }
            }
        }
        let target = parts.join("/");
        if unsafe_path
            || target.nfc().collect::<String>() != target
            || !self.entries.contains_key(&target)
        {
            self.fail(path, "SPD-SEC-008");
            return;
        }
        let Ok(fragment) = percent_decode_str(fragment).decode_utf8() else {
            self.fail(path, "SPD-SEC-008");
            return;
        };
        if !fragment.is_empty()
            && !self.documents.get(&target).is_some_and(|doc| {
                doc.descendants()
                    .any(|n| n.attribute("id") == Some(fragment.as_ref()))
            })
        {
            self.fail(path, "SPD-SEC-008");
        }
        if !link && let Some(source) = source {
            self.edges
                .entry(source)
                .or_default()
                .insert(if fragment.is_empty() {
                    target
                } else {
                    format!("{target}#{fragment}")
                });
        }
    }

    fn css(&mut self, path: &str, text: &str, svg: bool) {
        let mut input = ParserInput::new(text);
        let mut facts = CssFacts::default();
        tokens(&mut Parser::new(&mut input), "image", false, 0, &mut facts);
        if facts.code {
            self.fail(path, "SPD-SEC-001");
        }
        if facts.local_font {
            self.fail(path, "SPD-SEC-004");
        }
        if facts.malformed {
            self.fail(path, "SPD-SEC-005");
        }
        if facts.unresolved || facts.limited {
            self.issue(
                path,
                "SPD-SEC-008",
                if facts.limited {
                    "RESOURCE_LIMIT CSS nesting"
                } else {
                    "dynamic URL consumer not resolved"
                },
                Outcome::NotTested,
            );
        }
        for (url, role, import) in facts.urls {
            self.reference(
                path,
                &url,
                role,
                svg,
                false,
                import.then(|| path.to_owned()),
            );
        }
    }

    fn xml(&mut self, path: &str, doc: &Document<'_>) {
        for pi in doc
            .descendants()
            .filter_map(|n| n.pi())
            .filter(|p| p.target == "xml-stylesheet")
        {
            let wrapped = format!("<style {}/>", pi.value.unwrap_or(""));
            if let Ok(attrs) = Document::parse(&wrapped) {
                if let Some(href) = attrs.root_element().attribute("href") {
                    self.reference(path, href, "stylesheet", false, false, None);
                }
            } else {
                self.fail(path, "SPD-SEC-005");
            }
        }
        for n in doc.descendants().filter(|n| n.is_element()) {
            let ns = n.tag_name().namespace().unwrap_or("");
            if ![HTML, SVG, MATH].contains(&ns) {
                continue;
            }
            let name = n.tag_name().name().to_ascii_lowercase();
            let svg = n.ancestors().any(|a| a.tag_name().namespace() == Some(SVG));
            if name == "script" {
                self.fail(path, "SPD-SEC-001");
                if svg {
                    self.fail(path, "SPD-SEC-007");
                }
            }
            if ["iframe", "object", "embed", "applet"].contains(&name.as_str()) {
                self.fail(path, "SPD-SEC-002");
                self.fail(path, "SPD-PASS-002");
            }
            if ["form", "base", "canvas"].contains(&name.as_str())
                || n.attribute(("http://www.w3.org/XML/1998/namespace", "base"))
                    .is_some()
            {
                self.fail(path, "SPD-PASS-002");
            }
            if name == "foreignobject" && svg {
                self.fail(path, "SPD-SEC-007");
                self.fail(path, "SPD-PASS-002");
            }
            if ["input", "button"].contains(&name.as_str()) {
                let label = n
                    .attribute("aria-label")
                    .or(n.attribute("aria-labelledby"))
                    .is_some()
                    || (name == "button"
                        && n.descendants()
                            .any(|t| t.is_text() && !t.text().unwrap_or("").trim().is_empty()))
                    || n.attribute("id").is_some_and(|id| {
                        doc.descendants()
                            .any(|l| l.has_tag_name("label") && l.attribute("for") == Some(id))
                    });
                let bad_attr = n.attributes().any(|a| {
                    [
                        "form",
                        "formaction",
                        "formenctype",
                        "formmethod",
                        "formtarget",
                        "formnovalidate",
                        "command",
                        "commandfor",
                        "popovertarget",
                        "popovertargetaction",
                    ]
                    .contains(&a.name())
                });
                let kind = n
                    .attribute("type")
                    .unwrap_or(if name == "button" { "submit" } else { "text" })
                    .to_ascii_lowercase();
                if n.attribute("disabled").is_none()
                    || !label
                    || bad_attr
                    || ["password", "file", "submit", "image", "reset"].contains(&kind.as_str())
                    || (name == "button" && kind != "button")
                    || (name == "input"
                        && n.attribute("value").is_none()
                        && !["checkbox", "radio"].contains(&kind.as_str()))
                {
                    self.fail(path, "SPD-PASS-002");
                }
            }
            if name == "meta"
                && n.attribute("http-equiv")
                    .is_some_and(|v| v.eq_ignore_ascii_case("refresh"))
            {
                self.fail(path, "SPD-SEC-003");
                self.fail(path, "SPD-PASS-002");
            }
            if n.attribute("autoplay").is_some() {
                self.fail(path, "SPD-PASS-003");
            }
            let rel = n.attribute("rel").unwrap_or("").to_ascii_lowercase();
            if n.attribute("ping").is_some()
                || rel
                    .split_whitespace()
                    .any(|v| ["prefetch", "preconnect", "dns-prefetch", "prerender"].contains(&v))
            {
                self.fail(path, "SPD-SEC-003");
            }
            if name == "style" {
                let text: String = n
                    .descendants()
                    .filter(|v| v.is_text())
                    .filter_map(|v| v.text())
                    .collect();
                self.css(path, &text, svg);
            }
            for a in n.attributes() {
                let attr = a.name().to_ascii_lowercase();
                if attr.starts_with("on") {
                    self.fail(path, "SPD-SEC-001");
                    if svg {
                        self.fail(path, "SPD-SEC-007");
                    }
                }
                if attr == "style"
                    || (svg
                        && [
                            "fill",
                            "stroke",
                            "filter",
                            "clip-path",
                            "mask",
                            "cursor",
                            "marker-start",
                            "marker-mid",
                            "marker-end",
                        ]
                        .contains(&attr.as_str()))
                {
                    self.css(path, a.value(), svg);
                }
                if !["href", "src", "srcset", "poster", "data", "background"]
                    .contains(&attr.as_str())
                {
                    continue;
                }
                if name == "base" {
                    continue;
                }
                let link = attr == "href"
                    && (["a", "area"].contains(&name.as_str()) || (ns == MATH && name != "mglyph"));
                let mut role = if name == "link"
                    && rel.split_whitespace().any(|s| s == "stylesheet")
                {
                    "stylesheet"
                } else if [
                    "img", "image", "use", "feimage", "mglyph", "audio", "video", "source", "track",
                ]
                .contains(&name.as_str())
                    || ["poster", "background"].contains(&attr.as_str())
                {
                    "image"
                } else {
                    "other"
                };
                if name == "link"
                    && rel
                        .split_whitespace()
                        .any(|s| matches!(s, "preload" | "modulepreload"))
                {
                    role = match n.attribute("as") {
                        Some("font") => "font",
                        Some("style") => "stylesheet",
                        Some("image") => "image",
                        _ => role,
                    };
                }
                let values: Vec<&str> = if attr == "srcset" {
                    a.value()
                        .split(',')
                        .filter_map(|v| v.split_whitespace().next())
                        .collect()
                } else {
                    vec![a.value()]
                };
                for value in values {
                    let source = (svg && !link).then(|| {
                        n.ancestors()
                            .find_map(|a| a.attribute("id"))
                            .map(|id| format!("{path}#{id}"))
                            .unwrap_or_else(|| path.to_owned())
                    });
                    self.reference(path, value, role, svg, link, source);
                }
            }
        }
    }
}

pub fn inspect(entries: &BTreeMap<String, Entry>, findings: &mut Vec<Finding>) {
    let xml_text: BTreeMap<_, _> = entries
        .iter()
        .filter_map(|(p, e)| {
            let b = &e.bytes;
            let text = if b.starts_with(&[0xff, 0xfe]) || b.starts_with(b"<\0") {
                let start = if b.starts_with(&[0xff, 0xfe]) { 2 } else { 0 };
                if (b.len() - start) % 2 != 0 {
                    return None;
                }
                String::from_utf16(
                    &b[start..]
                        .as_chunks::<2>()
                        .0
                        .iter()
                        .map(|c| u16::from_le_bytes([c[0], c[1]]))
                        .collect::<Vec<_>>(),
                )
                .ok()?
            } else if b.starts_with(&[0xfe, 0xff]) || b.starts_with(b"\0<") {
                let start = if b.starts_with(&[0xfe, 0xff]) { 2 } else { 0 };
                if (b.len() - start) % 2 != 0 {
                    return None;
                }
                String::from_utf16(
                    &b[start..]
                        .as_chunks::<2>()
                        .0
                        .iter()
                        .map(|c| u16::from_be_bytes([c[0], c[1]]))
                        .collect::<Vec<_>>(),
                )
                .ok()?
            } else {
                std::str::from_utf8(b)
                    .ok()?
                    .trim_start_matches('\u{feff}')
                    .to_owned()
            };
            if !text.trim_start().starts_with('<') {
                return None;
            }
            Some((p.clone(), text))
        })
        .collect();
    let documents: BTreeMap<_, _> = xml_text
        .iter()
        .filter_map(|(p, text)| Document::parse(text).ok().map(|d| (p.clone(), d)))
        .collect();
    let mut scan = Inspection {
        entries,
        documents,
        findings,
        edges: BTreeMap::new(),
    };
    // Separate parse handles avoid mutable/immutable aliasing during inspection.
    let paths: Vec<_> = scan.documents.keys().cloned().collect();
    let mut css_paths: BTreeSet<String> = entries
        .keys()
        .filter(|p| p.ends_with(".css"))
        .cloned()
        .collect();
    for path in paths {
        if let Some(text) = xml_text.get(&path)
            && let Ok(doc) = Document::parse(text)
        {
            if doc.root_element().tag_name().namespace() == Some("http://www.idpf.org/2007/opf") {
                for item in doc.descendants().filter(|n| n.has_tag_name("item")) {
                    if item.attribute("media-type").is_some_and(|m| {
                        matches!(
                            m,
                            "text/javascript"
                                | "application/javascript"
                                | "text/ecmascript"
                                | "application/ecmascript"
                        )
                    }) {
                        scan.fail(&path, "SPD-SEC-001");
                    }
                }
                for item in doc
                    .descendants()
                    .filter(|n| n.attribute("media-type") == Some("text/css"))
                {
                    if let Some(href) = item.attribute("href")
                        && !href.contains(':')
                    {
                        let base = path
                            .rsplit_once('/')
                            .map(|(p, _)| format!("{p}/"))
                            .unwrap_or_default();
                        css_paths.insert(format!("{base}{href}"));
                    }
                }
            }
            scan.xml(&path, &doc);
        }
    }
    for path in css_paths {
        if let Some(entry) = entries.get(&path) {
            if let Ok(text) = std::str::from_utf8(&entry.bytes) {
                scan.css(&path, text.trim_start_matches('\u{feff}'), false);
            } else {
                scan.fail(&path, "SPD-SEC-005");
            }
        }
    }
    let mut done = BTreeSet::new();
    let roots: Vec<_> = scan.edges.keys().cloned().collect();
    for root in roots {
        let mut active = BTreeSet::new();
        let mut stack = vec![(root.clone(), false)];
        while let Some((node, leave)) = stack.pop() {
            if leave {
                active.remove(&node);
                done.insert(node);
            } else if active.contains(&node) {
                scan.fail(root.split('#').next().unwrap_or(&root), "SPD-SEC-008");
            } else if !done.contains(&node) {
                active.insert(node.clone());
                stack.push((node.clone(), true));
                if let Some(children) = scan.edges.get(&node) {
                    stack.extend(children.iter().map(|v| (v.clone(), false)));
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    fn css(text: &str) -> CssFacts {
        let mut source = ParserInput::new(text);
        let mut result = CssFacts::default();
        tokens(
            &mut Parser::new(&mut source),
            "image",
            false,
            0,
            &mut result,
        );
        result
    }
    #[test]
    fn nested_escaped_resource_consumers_and_inert_strings() {
        let parsed = css(
            r#"@media screen { a { --x: u\72l(https://example.invalid/x); background:var(--x); } }"#,
        );
        assert_eq!(parsed.urls[0].0, "https://example.invalid/x");
        assert!(
            css(r#"p::after {content:"https://example.invalid/inert"}"#)
                .urls
                .is_empty()
        );
        assert_eq!(css(r#"@import "local.css" print;"#).urls[0].1, "stylesheet");
        assert_eq!(
            css(r#"@font-face { src: url(font.ttf) }"#).urls[0].1,
            "font"
        );
        assert!(css(r#"p {background:image-set(var(--source) 1x)}"#).unresolved);
    }
    #[test]
    fn namespace_context_does_not_hide_active_svg() {
        let entries = BTreeMap::from([("no-extension".into(), Entry { bytes: br#"<svg:svg xmlns:svg="http://www.w3.org/2000/svg"><svg:script>void(0)</svg:script></svg:svg>"#.to_vec(), size: 0, sha256: String::new() })]);
        let mut findings = Vec::new();
        inspect(&entries, &mut findings);
        for id in ["SPD-SEC-001", "SPD-SEC-007"] {
            assert!(findings.iter().any(|f| f.requirement_id == id));
        }
    }
    proptest::proptest! {
        #[test]
        fn arbitrary_css_is_bounded_and_does_not_panic(text in ".{0,4096}") { let _ = css(&text); }
    }
}
